### PR TITLE
HTML5 fixes, refactor, audio fallback, fixed FPS.

### LINKED ISSUE #38623: Low processor usage mode on HTML5 exports
**Godot version:**
v.3.2.1

**OS/device including version:**
Brave on Mac
Safari on Mac

**Issue description:**
In playing around with Godot I created a test app that is essentially just a series of UI controls and deployed it to the web. In trying to reduce battery usage I came across the "low processor usage mode", which sounded like a good candidate. In testing it, however, it seems to result in more work being done when enabled than when it's disabled. It almost looks like the complete opposite to what I think I'd expect.

Some example profiles I've saved - these can be imported into the performance tab of developer tools for Brave (and presumably anything chrome-y): [Archive.zip](https://github.com/godotengine/godot/files/4605542/Archive.zip)

**Screenshots**
Brave - Enabled
Screen Shot 2020-05-10 at 12 17 40
Safari - Enabled
Screen Shot 2020-05-10 at 12 22 24

Brave - Disabled
Screen Shot 2020-05-10 at 12 17 43
Safari - Disabled
Screen Shot 2020-05-10 at 12 21 32

**Steps to reproduce:**
1. Clone the project below.
2. Run it as an HTML5 export with low processor usage mode disabled via project settings.
3. Go to performance tab of developer tools and profile for 30 seconds.
4. Enable low processor usage mode and re-run as an HTML5 export.
5. Go to performance tab of developer tools and profile for 30 seconds again.
6. Compare and see the (significantly) higher scripting usage with low processor mode enabled.
 
**Minimal reproduction project:**
https://github.com/samuelhwilliams/test-low-processor-usage

### LINKED ISSUE #39930: Slow/Stutter iOS html5 Audio fix in 3.2.2 broke audio in android
**Godot version:**
3.2.2 Stable

**OS/device including version:**
Android 8/9

**Issue description:**
Audio (ogg, 44.1) Plays super chopped and slow on android devices.
https://github.com/godotengine/godot/issues/35937 

**Steps to reproduce:**
AudioStreamPlayer inside 2D Node

### PR DESCRIPTION
**Pull request by Faless**
In this PR:

- Fix for HTML5 startup FS error (after #39563, referenced for cherry-pick).
- Refactor `DisplayServerJavaScript` to better handle canvas ID and avoid polluting global namespace.
- Canvas size check on each loop to force redrawing when the size changed (tested in a 3.2 backport).
- Working `Engine.target_fps` in HTML5 without blocking the main thread (fixes #38623).
- Fallback to `AudioDriverDummy` when `window.AudioContext` is not supported by the browser.
- Higher (50 ms) default `audio/output_latency` override for web export (fixes #39930).

I'll make another PR for the `3.2` since cherry-picking is not that easy.

### CODE REVIEW

File: platform/javascript/audio_driver_javascript.cpp [hunk 1]
```diff
@@ -36,6 +36,15 @@
 
 AudioDriverJavaScript *AudioDriverJavaScript::singleton = nullptr;
 
+bool AudioDriverJavaScript::is_available() {
+	return EM_ASM_INT({
```
Comment (akien-mga) [inline platform/javascript/audio_driver_javascript.cpp:40]: Doesn't this need an explicit `(bool)` cast? (If there isn't an `EM_ASM_BOOL`).

Comment (akien-mga) [reply platform/javascript/audio_driver_javascript.cpp:40]: (Unrelated, but you lost the end of the commit title :P)

File: platform/javascript/os_javascript.cpp [hunk 2]
```diff
@@ -238,7 +242,10 @@ void OS_JavaScript::initialize_joypads() {
 }
 
 OS_JavaScript::OS_JavaScript() {
-	AudioDriverManager::add_driver(&audio_driver_javascript);
+	if (AudioDriverJavaScript::is_available()) {
+		audio_driver_javascript = memnew(AudioDriverJavaScript);
```
Comment (akien-mga) [inline platform/javascript/os_javascript.cpp:246]: Shouldn't it be `memdelete()`-ed?

File: platform/javascript/javascript_main.cpp [hunk 3]
```diff
@@ -47,9 +48,18 @@ void exit_callback() {
 }
 
 void main_loop_callback() {
+	uint64_t current_ticks = os->get_ticks_usec();
+
 	bool force_draw = DisplayServerJavaScript::get_singleton()->check_size_force_redraw();
 	if (force_draw) {
 		Main::force_redraw();
+	} else if (current_ticks < target_ticks && !force_draw) {
```
Comment (akien-mga) [inline platform/javascript/javascript_main.cpp:56]: The `&& !force_draw` part is unnecessary, you can't enter this `else if` condition if `force_draw` was true.

File: platform/javascript/os_javascript.h [hunk 4]
```diff
@@ -83,6 +83,7 @@ class OS_JavaScript : public OS_Unix {
 	String get_executable_path() const;
 	virtual Error shell_open(String p_uri);
 	virtual String get_name() const;
+	virtual void add_frame_delay(bool p_can_draw) {}
```
Comment (akien-mga) [inline platform/javascript/os_javascript.h:86]: Might use a comment explaining that it's define as no-op to override the base `OS` implementation, which is redone in `main_loop_callback()`.

File: platform/javascript/display_server_javascript.cpp [hunk 5]
```diff
@@ -45,21 +45,22 @@
...
 char DisplayServerJavaScript::canvas_id[256] = { 0 };
+static bool cursor_inside_canvas = true;
 
 DisplayServerJavaScript *DisplayServerJavaScript::get_singleton() {
 	return static_cast<DisplayServerJavaScript *>(DisplayServer::get_singleton());
 }
 
 // Window (canvas)
-static void focus_canvas() {
+void DisplayServerJavaScript::focus_canvas() {
```
Comment (akien-mga) [inline platform/javascript/display_server_javascript.cpp:55]: The commit says "More", but it's "Less", no? ;)
(Also typo in "Where" but that's really nitpicking.)

**Comment by akien-mga** [comment]
Done with my review, left some comments but looks great overall.

File: platform/javascript/audio_driver_javascript.cpp (hunk 1 above)
Comment (Faless) [reply platform/javascript/audio_driver_javascript.cpp:40]: Well, there is no `EM_ASM_BOOL`, and int to bool seem safe in this case, but I appended `!= 0` at the end so the function returns a boolean.

File: platform/javascript/os_javascript.cpp (hunk 2 above)
Comment (Faless) [reply platform/javascript/os_javascript.cpp:250]: Yes! Thanks for spotting it.

File: platform/javascript/javascript_main.cpp (hunk 3 above)
Comment (Faless) [reply platform/javascript/javascript_main.cpp]: Yep, so much rebasing ;). Ty

File: platform/javascript/os_javascript.h (hunk 4 above)
Comment (Faless) [reply platform/javascript/os_javascript.h:88]: Done

File: platform/javascript/display_server_javascript.cpp (hunk 5 above)
Comment (Faless) [reply platform/javascript/display_server_javascript.cpp:55]: The rest of the commit message should explain it: `Were static methods in cpp file, polluting global namespace`.
I.e. they were static functions, they are now static methods inside the class DisplayServerJavaScript.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**